File: packages/python/src/kingpepe_sdk/money.py
# Copyright (c) 2026 The KingPepe developers
# Distributed under the MIT software license, see the accompanying LICENSE file.
"""Money handling for KPEPE.

KPEPE amounts are ALWAYS represented internally as integer base units. Binary
floating point (``float``) is never used for value-carrying amounts. Accept a
validated decimal string or an int of base units; never a float.
"""

from __future__ import annotations

import re
from decimal import Decimal, InvalidOperation

from .errors import ValidationError
from .network import BASE_UNITS_PER_KPEPE, KPEPE_DECIMALS, MAX_MONEY_BASE_UNITS

_AMOUNT_RE = re.compile(r"^-?\d+(\.\d+)?$")


def parse_kpepe(amount: str) -> int:
    """Parse a decimal KPEPE string (e.g. ``"1.23456789"``) into integer base units."""
    if not isinstance(amount, str):
        raise ValidationError("Amount must be a decimal string, not a float (avoid precision loss).")
    text = amount.strip()
    if not _AMOUNT_RE.match(text):
        raise ValidationError(f"Invalid KPEPE amount: {amount!r}")
    negative = text.startswith("-")
    unsigned = text[1:] if negative else text
    whole, _, frac = unsigned.partition(".")
    if len(frac) > KPEPE_DECIMALS:
        raise ValidationError(f"KPEPE supports at most {KPEPE_DECIMALS} decimals; got {len(frac)}.")
    frac_padded = frac.ljust(KPEPE_DECIMALS, "0")
    base = int(whole) * BASE_UNITS_PER_KPEPE + (int(frac_padded) if frac_padded else 0)
    value = -base if negative else base
    _assert_money_range(value)
    return value


def format_kpepe(base_units: int) -> str:
    """Format integer base units as a fixed 8-decimal KPEPE string."""
    if not isinstance(base_units, int) or isinstance(base_units, bool):
        raise ValidationError("Base units must be an int.")
    negative = base_units < 0
    abs_units = -base_units if negative else base_units
    whole, frac = divmod(abs_units, BASE_UNITS_PER_KPEPE)
    return f"{'-' if negative else ''}{whole}.{frac:0{KPEPE_DECIMALS}d}"


#: Aliases matching the documented conversion helpers.
kpepe_to_base_units = parse_kpepe
base_units_to_kpepe = format_kpepe


def validate_amount(amount: str | int) -> int:
    """Validate a decimal string or base-unit int is a non-negative, in-range value."""
    base_units = amount if isinstance(amount, int) and not isinstance(amount, bool) else parse_kpepe(str(amount))
    if base_units < 0:
        raise ValidationError("Amount must not be negative.")
    _assert_money_range(base_units)
    return base_units


def rpc_amount_to_base_units(amount: float | int | str) -> int:
    """Convert a KPEPE amount returned by Core RPC into exact integer base units.

    Core reports amounts as JSON numbers with up to 8 decimals; using Decimal on
    the repr keeps the conversion exact within the money range.
    """
    try:
        dec = Decimal(str(amount))
    except InvalidOperation as exc:  # pragma: no cover - defensive
        raise ValidationError(f"Invalid RPC amount: {amount!r}") from exc
    scaled = (dec * BASE_UNITS_PER_KPEPE).quantize(Decimal(1))
    base_units = int(scaled)
    _assert_money_range(base_units)
    return base_units


def _assert_money_range(base_units: int) -> None:
    if abs(base_units) > MAX_MONEY_BASE_UNITS:
        raise ValidationError("Amount out of range (exceeds MAX_MONEY).")

File: packages/javascript/src/money.ts
// Copyright (c) 2026 The KingPepe developers
// Distributed under the MIT software license, see the accompanying LICENSE file.

import { BASE_UNITS_PER_KPEPE, KPEPE_DECIMALS, MAX_MONEY_BASE_UNITS } from "./network.js";
import { ValidationError } from "./errors.js";

/**
 * Money handling for KPEPE.
 *
 * KPEPE amounts are ALWAYS represented internally as integer base units
 * (bigint). Binary floating point (`number`) is never used for value-carrying
 * amounts, because it cannot represent decimal fractions exactly. Accept either
 * a validated decimal string or a bigint of base units; never a float.
 */

const AMOUNT_RE = /^-?\d+(\.\d+)?$/;

/**
 * Parse a decimal KPEPE string (e.g. "1.23456789") into integer base units.
 * Rejects floats, more than 8 decimals, and out-of-range values.
 */
export function parseKpepe(amount: string): bigint {
  if (typeof amount !== "string") {
    throw new ValidationError("Amount must be a decimal string, not a number (avoid float precision loss).");
  }
  const trimmed = amount.trim();
  if (!AMOUNT_RE.test(trimmed)) {
    throw new ValidationError(`Invalid KPEPE amount: ${JSON.stringify(amount)}`);
  }
  const negative = trimmed.startsWith("-");
  const unsigned = negative ? trimmed.slice(1) : trimmed;
  const [wholePart = "0", fracPartRaw = ""] = unsigned.split(".");
  if (fracPartRaw.length > KPEPE_DECIMALS) {
    throw new ValidationError(`KPEPE supports at most ${KPEPE_DECIMALS} decimal places; got ${fracPartRaw.length}.`);
  }
  const fracPart = fracPartRaw.padEnd(KPEPE_DECIMALS, "0");
  const base = BigInt(wholePart) * BASE_UNITS_PER_KPEPE + BigInt(fracPart || "0");
  const signed = negative ? -base : base;
  assertMoneyRange(signed);
  return signed;
}

/** Format integer base units as a fixed 8-decimal KPEPE string (e.g. "1.23456789"). */
export function formatKpepe(baseUnits: bigint): string {
  if (typeof baseUnits !== "bigint") {
    throw new ValidationError("Base units must be a bigint.");
  }
  const negative = baseUnits < 0n;
  const abs = negative ? -baseUnits : baseUnits;
  const whole = abs / BASE_UNITS_PER_KPEPE;
  const frac = abs % BASE_UNITS_PER_KPEPE;
  const fracStr = frac.toString().padStart(KPEPE_DECIMALS, "0");
  return `${negative ? "-" : ""}${whole.toString()}.${fracStr}`;
}

/** Convert whole/decimal KPEPE (string) to base units. Alias of {@link parseKpepe}. */
export const kpepeToBaseUnits = parseKpepe;

/** Convert base units (bigint) to a KPEPE decimal string. Alias of {@link formatKpepe}. */
export const baseUnitsToKpepe = formatKpepe;

/**
 * Validate that an amount (decimal string or base-unit bigint) is a well-formed,
 * in-range, non-negative KPEPE value. Returns the amount in base units.
 */
export function validateAmount(amount: string | bigint): bigint {
  const baseUnits = typeof amount === "bigint" ? amount : parseKpepe(amount);
  if (baseUnits < 0n) {
    throw new ValidationError("Amount must not be negative.");
  }
  assertMoneyRange(baseUnits);
  return baseUnits;
}

/**
 * Convert base units to a JS number of KPEPE. Provided only for display/among
 * non-value-critical contexts; NEVER use the result to move funds.
 */
export function baseUnitsToKpepeNumberUnsafe(baseUnits: bigint): number {
  return Number(formatKpepe(baseUnits));
}

/**
 * Convert a KPEPE amount as returned by Core RPC (a JSON number) into exact
 * integer base units. Core's wire format for amounts is a decimal number with
 * up to 8 places; every in-range base-unit integer (<= 2.1e15) is exactly
 * representable as a double, so rounding the scaled value is lossless here.
 */
export function rpcAmountToBaseUnits(amount: number): bigint {
  if (!Number.isFinite(amount)) {
    throw new ValidationError(`Invalid RPC amount: ${amount}`);
  }
  const scaled = Math.round(amount * Number(BASE_UNITS_PER_KPEPE));
  const baseUnits = BigInt(scaled);
  assertMoneyRange(baseUnits);
  return baseUnits;
}

function assertMoneyRange(baseUnits: bigint): void {
  const abs = baseUnits < 0n ? -baseUnits : baseUnits;
  if (abs > MAX_MONEY_BASE_UNITS) {
    throw new ValidationError("Amount out of range (exceeds MAX_MONEY).");
  }
}
